cmd/snap/cmd_run.go

```diff
@@ -118,6 +118,40 @@ func getSnapInfo(snapName string, revision snap.Revision) (*snap.Info, error) {
 	return info, nil
 }
 
+func createOrUpdateUserDataSymlink(info *snap.Info, usr *user.User) error {
+	// create /home/snap/$SNAP_NAME/.lock file
+	lockFilePath := filepath.Join(info.HomeDirBase(usr.HomeDir), ".lock")
+	lockFd, err := syscall.Open(lockFilePath, syscall.O_CREAT, syscall.S_IRUSR|syscall.S_IWUSR)
+	if err != nil {
+		// TRANSLATORS: %q is the file path, %v the error message
+		return fmt.Errorf(i18n.G("cannot create lock file %q: %v"), lockFilePath, err)
+	}
+	err = syscall.Flock(lockFd, syscall.LOCK_EX)
```

Comment:
Why do we need the lock at all?  Why isn't this done opportunistically on a first come first serve basis?  It's okay to attempt to do it twice. Both processes should arrive at the same conclusion. Do it in a loop, and rerun for a defined number of times if concurrency is detected (say, 3?) until the expected outcome is achieved.